sdk/storage/Azure.Storage.Common/src/Shared/WriteStream.cs

```diff
@@ -0,0 +1,181 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+namespace Azure.Storage.Shared
+{
+    internal abstract class WriteStream : Stream
+    {
+        protected long _position;
+        protected readonly IProgress<long> _progressHandler;
+        protected readonly MemoryStream _buffer;
+
+        protected WriteStream(
+            long position,
+            int bufferSize,
+            IProgress<long> progressHandler)
+        {
+            _position = position;
+            _progressHandler = progressHandler;
+            _buffer = new MemoryStream(bufferSize);
+        }
+
+        public override bool CanRead => false;
+
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => true;
+
+        public override long Length => throw new NotSupportedException();
+
+        public override long Position { get => _position; set => throw new NotSupportedException(); }
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        public override long Seek(long offset, SeekOrigin origin)
+            => throw new NotSupportedException();
+
+        public override void SetLength(long value)
+            => throw new NotSupportedException();
+
+        public override void Write(
+            byte[] buffer,
+            int offset,
+            int count)
+            => WriteInternal(
+                async: false,
+                buffer,
+                offset,
+                count,
+                cancellationToken: default)
+            .EnsureCompleted();
+
+        public override async Task WriteAsync(
+            byte[] buffer,
+            int offset,
+            int count,
+            CancellationToken cancellationToken)
+            => await WriteInternal(
+                async: true,
+                buffer,
+                offset,
+                count,
+                cancellationToken)
+            .ConfigureAwait(false);
+
+        protected virtual async Task WriteInternal(
+            bool async,
+            byte[] buffer,
+            int offset,
+            int count,
+            CancellationToken cancellationToken)
+        {
+            ValidateWriteParameters(buffer, offset, count);
+            int remaining = count;
+
+            // New bytes will fit in the buffer.
+            if (count <= _buffer.Capacity - _buffer.Position)
+            {
+                await WriteToBuffer(async, buffer, offset, count, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                // Flush the buffer.
+                await FlushInternal(async, cancellationToken).ConfigureAwait(false);
```

Comment:
should we write as much as we can into remaining buffer before flushing? Otherwise we're not filling blocks most optimal way (assuming that there will be cases hitting block count limit at some point).

Reply:
Did this for simplicity, I'm working to modify it to fill the buffer first.  Also, I believe this is supposed to be AppendInternal call.